java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventData.java

```diff
@@ -196,43 +207,130 @@ public EventData(ByteBuffer buffer)
 	/**
 	 * Get Actual Payload/Data wrapped by EventData.
 	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
-	 * @return returns the byte[] of the actual data 
+	 * @return returns the byte[] of the actual data
+	 * @deprecated As of version 0.8.0. Use {@link #getPayloadArray()} {@link #getPayloadArrayOffset()} {@link #getPayloadArrayLength()}
 	 */
-	public byte[] getBody()
+	@Deprecated public byte[] getBody()
 	{
-		// TODO: enforce on-send constructor type 2
 		return this.bodyData == null ? null : this.bodyData.getArray();
 	}
+	
+	/**
+	 * Get Actual Payload/Data wrapped by EventData.
+	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
+	 * @return returns the byte[] of the actual data
+	 * @see #getPayloadArrayLength()
+	 * @see #getPayloadArrayOffset()
+	 */
+	public byte[] getPayloadArray()
+	{
+		return this.bodyData == null ? null : this.bodyData.getArray();
+	}
+	
+	/**
+	 * Get Actual Payload/Data wrapped by EventData.
+	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
+	 * @return returns the byte[] of the actual data
+	 * @see #getPayloadArrayLength()
+	 * @see #getPayloadArray()
+	 */
+	public int getPayloadArrayOffset()
+	{
+		return this.bodyData == null ? 0 : this.bodyData.getArrayOffset();
+	}
+	
+	/**
+	 * Get Actual Payload/Data wrapped by EventData.
+	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
+	 * @return returns the byte[] of the actual data
+	 * @see #getPayloadArray()
+	 * @see #getPayloadArrayOffset()
+	 */
+	public int getPayloadArrayLength()
+	{
+		return this.bodyData == null ? 0 : this.bodyData.getLength();
+	}
 
 	/**
 	 * Application property bag
 	 * @return returns Application properties
 	 */
 	public Map<String, String> getProperties()
 	{
-		return this.properties;
-	}
+		if (this.properties == null)
```

Comment:
>if (this.properties == null) [](start = 2, length = 28)

do we need to account for multi-thread usage here? we never say we are not thread safe in our library. so if we assume user can call this in multiple threads we need a sync block here.